LSPS5/README.md

```diff
@@ -0,0 +1,264 @@
+  device as a `GET` parameter.
+
+> **Rationale** By placing any data to the `GET` parameters
+> of the given URI, the LSP interface is simplified and
+> the client developer has the flexibility to add any
+> additional information they want the LSP to simply
+> pass through from the client instance to the client
+> application server.
+
+The LSP:
+
+* MUST persist the given webhook and associate it with the
+  client node ID before responding to this call.
+* MUST use the `webhook` specified in the latest
+  `lsps5.setwebhook` call.
+* MAY forget the `webhook` after 7 days (7 x 24 hours) from
+  the definite closure of the last channel between the LSP
+  and the client.
+  * MUST treat a possible channel open, such as via a
+    reserved SCID for a JIT channel, as a channel open, for
+    purposes of remembering the `webhook` for a client.
+
+The `lsps5.setwebhook` call results in the empty object
+`{}` on success, and has the following errors defined:
+
+* `unsupported_version` (1) - The `version` specified by
+  the client is not among the `versions` that the LSP
+  supports.
+* `unsupported_protocol` (2) - The `webhook` is not an
+  HTTPS-protocol URI.
+
+### Webhook Call
+
+The LSP SHOULD make a `POST` request to the given webhook
+URI, if the client has registered one via `lsps5.setwebhook`
+*and* one of the following conditions is true:
+
+* The LSP learns of an incoming payment to the client
+  (whether an incoming HTLC, or other future mechanism
+  for incoming payments).
+* The LSP wants to take back some of its liquidity towards
+  the client.
+* The LSP wants to change feerates for any channels towards
+  the client, or feerates for any auto-management mechanism
+  the client has set up with the LSP.
+* The LSP wants to withdraw all services from the client
+  and close all channels with the client.
+
+The LSP makes a `POST` request whose content body has the
+MIME type `application/json`, containing a JSON object like
+the below:
+
+```JSON
+{
+  "version": 1,
+  "lsp": "023456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef012",
+  "details": {
+    "reason": 1,
+    "pushed_at": "2023-04-25T14:27:03.842Z",
+    "zbase": "blahblah"
+  }
+}
+```
+
+`version` is a required integral number, the version of the
+API that was specified in the `lsps5.setwebhook` call.
+
+`lsp` is a required node ID, the node ID of the LSP that is
+making this push request.
+
+`details` is a required object whose format is dependent on
+`version`.
+
+* For `version` 1, `details` has the fields:
+  * `reason` is a required integral number, which is one
+    of the reason codes defined below.
+  * `pushed_at` is a required datetime, which is the time
+    at the LSP when it generated this push request.
+  * `zbase` is a required signature from the LSP, signing
+    the message defined below.
+
+For `version` 1 `zbase` signatures, the signature signs
+the following JSON string:
+
+    "LSPS5:Version 1: Reason ${reason} at time ${pushed_at}."
```

Comment:
Can we generalize the signing? Here is an example. It signs the whole request body, url + timestamp. The receiver is able to validate the origin of the request generically without any custom strings like `"LSPS5:Version 1: Reason ${reason} at time ${pushed_at}."`. 

```js
import crypto from "crypto"
import axios from "axios"

const apiSecret = 'apisec_****'

const data = { amount: 1000 }
const body = JSON.stringify(data)

const method = 'POST' // Uppercase
const path = '/webhook'
const query = '?myQuery=123'

const timestamp = Date.now().valueOf().toString()
const msgToSign = `${timestamp}${method}${path}${query}${body}`
const signature = crypto.createHmac('sha256', apiSecret).update(msgToSign).digest('base64')
const headers = {
    'x-api-signature': signature,
    'x-api-timestamp': timestamp,
}

// make the HTTPS request with the above headers
axios.post('https://example.com/ + path, data, {
    headers
})
```

Some modifications are needed of course:

* We need to adapt it to use LN signatures. 
* We need to put the LSP pubkey in the header too to validate the signature. This also gives us free authentication.

Another advantage of this: Replay attacks can easily be prevented:

1. Check `x-api-timestamp` not to be off more than x min (for example 5min).
2. Cache `x-api-signature` for 2*x min (10min in this example) and decline any request with the same signature within this timeframe.

Signing like this is successfully used on production systems. For example: https://docs.bitfinex.com/docs/rest-auth.

Reply:
Hmmm.  Is this now common over in HTTP-land?

Reply:
Yes

Reply:
Okay so I thought about this a bit more and I think we should still retain the string template and not just do concatenation of the various parts, if we are going to use the `signmessage` protocol.

The issue is that the `signmessage` thing is also a human-accessible interface, and there are often informal bespoke protocols where someone will ask a public node operator to sign some message as evidence that they control a particular public node.

It is best if the message we pass to `signmessage` command includes "LSPS5" and probably things like "I will push a notification at a client", in human-readable text.  Otherwise someone using such an informal bespoke protocol could pass off what looks like gobbledegook to a human to generate an LSPS5 webhook authorizing signature (this probably ought to get mentioned over in common schemas, for all `signmessage` signatures; "Other LSPS specifications MUST include human-readable text in the message to sign, describing what is being committed to." and include this rationale).

Reply:
Then just do the following:

```js
const msgToSign = `LSPS5. THIS SHOULD NEVER BE SIGNED MANUALLY! For more information about LSPS5 see https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/38${timestamp}${method}${path}${query}${body}`
```

There is no reason to use custom signing.

Reply:
Umm yes that is what I said, we still need to have it in the template that is signed?

Reply:
@SeverinAlexB: @JssDWt writes in https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/38#discussion_r1181488576:

>I think I prefer the signature in the message itself. Just so it's easily portable to other protocols if necessary (like noise maybe?)

My own opinion is that if we want to use a different protocol other than HTTPS we should define a different version of LSPS5 and describe how to send the LSP node ID, current time, and signature.  For HTTPS we go with HTTP headers.  What do you think?

Reply:
That sounds reasonable to me.

Reply:
>That sounds reasonable to me.

Which one?  @JssDWt proposal (signature in the message body), or mine (signature out of message body, we define it per protocol how it is sent)?

Reply:
ahh sorry. I like your proposal (@ZmnSCPxj-jr) separating it into a new spec and keeping this spec http only.

But again, there is no other protocol than http needed anyway IMO.

Reply:
Went with specifying `x-api-signature` and `x-api-timestamp` HTTP headers for HTTPS webhooks.